### PR TITLE
MM-41211: Adds ability to set SessionLengthSSOInDays to a decimal.

### PR DESCRIPTION
**Pull request by mkraft**
#### Summary

* Adds ability to set SessionLengthSSOInDays to a decimal.
* Changes the names of some ambiguous constants.
* Changes some methods to operate on Hours rather than Days time units.
* Validates that the session expiry converts to at least 1 hour.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-41211

#### Related PRs

* https://github.com/mattermost/mattermost-webapp/pull/9686

#### Release Note

```release-note
Adds the ability to set SSO session length to a decimal (ex. `0.5` for 12 hours session lengths). Note to developer: code that uses the (model.ServiceSettings).SessionLengthSSOInDays field will need to be updated to handle the data type change from int to a float32.
```

### CODE REVIEW

**Review by hanzei — COMMENTED** [review]
I'm wondering the change would be clear (both in terms of code changes and in the config API), if we add a new config field instead of re-using an existing one.

File: model/config.go [hunk 1]
```diff
@@ -320,7 +320,7 @@ type ServiceSettings struct {
 	ExtendSessionLengthWithActivity                   *bool    `access:"environment_session_lengths,write_restrictable,cloud_restrictable"`
 	SessionLengthWebInDays                            *int     `access:"environment_session_lengths,write_restrictable,cloud_restrictable"`
 	SessionLengthMobileInDays                         *int     `access:"environment_session_lengths,write_restrictable,cloud_restrictable"`
-	SessionLengthSSOInDays                            *int     `access:"environment_session_lengths,write_restrictable,cloud_restrictable"`
+	SessionLengthSSOInDays                            *float32 `access:"environment_session_lengths,write_restrictable,cloud_restrictable"`
```
[the remark this answers is not in the document]
Comment (mkraft) [reply model/config.go:323]: The code snippet you've linked to is making the data type change at _runtime_. When a Mattermost instance is upgraded the process is stopped and at that point the only thing remaining is the `ServiceSettings.SessionLengthSSOInDays` value in JSON (ex `1` or `0.5` or whatever). When the app is restarted with the new code, all it knows is the new data type into which it parses the JSON value.

File: model/config.go [hunk 2]
```diff
@@ -3584,6 +3596,10 @@ func (s *ServiceSettings) isValid() *AppError {
 		return NewAppError("Config.IsValid", "model.config.is_valid.collapsed_threads.app_error", nil, "", http.StatusBadRequest)
 	}
 
+	if *s.SessionLengthSSOInHours() < 1 {
+		return NewAppError("Config.IsValid", "model.config.is_valid.session_length_sso_in_days.app_error", nil, "", http.StatusBadRequest)
+	}
```
[the remark this answers is not in the document]
Comment (mkraft) [reply model/config.go:3601]: Which other session length?

**Comment by mkraft** [comment]
> I'm wondering the change would be clear (both in terms of code changes and in the config API), if we add a new config field instead of re-using an existing one.

What would be the benefit of this approach? I couldn't think of one when I was considering this originally.

File: i18n/en.json [hunk 3]
```diff
@@ -8255,6 +8255,10 @@
     "id": "model.config.is_valid.saml_username_attribute.app_error",
     "translation": "Invalid Username attribute. Must be set."
   },
+  {
+    "id": "model.config.is_valid.session_length_sso_in_days.app_error",
+    "translation": " "
```
[the remark this answers is not in the document]
Comment (hanzei) [reply i18n/en.json:8260]: Thanks for the explanation :+1:

File: model/config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply model/config.go:3601]: `SessionLengthMobileInHours()` and `SessionLengthWebInHours()`. Or is that already check somewhere else?

File: model/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply model/config.go:323]: Sorry, my comment was ambiguous. My point is about other go code importing `model`, e.g. plugin or other integrations.

[the remark this answers is not in the document]
Comment (jespino) [reply model/config.go:323]: Yes, that is no longer compatible from "binary" perspective, I think it has very low risk, but still it is a backward compatibility problem from the go perspective. Still, to make it fail, I think you should need to explicitly use that configuration parameter in your source code. Can happen but is low risk. @agnivade what do you think about this?

**Review by jespino — APPROVED** [review]
It looks good to me except for the already raised concern about backward compatibility. If @agnivade is ok with that, we can move forward with this (IMO).

File: model/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (amyblais) [reply model/config.go:323]: @mkraft If we go ahead with the breaking change, please add a note to this PR's Release Note section about the breaking change.

File: model/config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mkraft) [reply model/config.go:3601]: The changes in this PR shouldn't be the cause for any new validations to those associated fields (`SessionLengthWebInDays` or `SessionLengthMobileInDays`) unless those have their data types changed in the future.

File: model/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mkraft) [reply model/config.go:323]: @amyblais Yes I did, you can see it if you scroll right.

**Review by hanzei — APPROVED** [review]

File: model/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mkraft) [reply model/config.go:323]: @hanzei @jespino @agnivade Are you proposing that the entire model adhere to semver or just the config?

**Pull request merged**